### PR TITLE
refactor(hooks): consolidate PostToolUse hooks into sync/async dispatchers

### PR DESCRIPTION
**Pull request by pythonstrup**
## What Changed

Consolidated the 10 individual `PostToolUse` entries in `hooks/hooks.json` into two dispatcher entries:

- `post:dispatcher:sync` — runs the 7 blocking hooks (design-quality-check, accumulator, console-warn, governance-capture, session-activity-tracker, metrics-bridge, context-monitor) in one process
- `post:dispatcher:async` — runs the 3 background hooks (bash dispatcher, quality-gate, continuous-learning observe) in one backgrounded process (`async: true`, timeout 45s)

The dispatcher's internal registry (`scripts/hooks/posttooluse-dispatcher.js`) preserves every original hook ID, matcher, and profile, so `ECC_DISABLED_HOOKS` and `ECC_HOOK_PROFILE` gating behave exactly as before.

Supporting changes:
- `scripts/dashboard-web.js` lists dispatcher-managed child hooks so the hook inventory stays complete
- `scripts/hooks/post-edit-console-warn.js` refactored to export `run()` for in-process dispatch while keeping standalone stdin behavior
- Dispatcher stdin reading is multi-byte safe (`StringDecoder`), truncation suppresses pass-through, and child hook exit codes propagate
- `commands/quality-gate.md` docs updated to describe the new wiring

## Why This Change

Every tool call spawned up to 10 Node processes (7 of them blocking the tool result), each paying full Node startup plus the plugin-root bootstrap. Consolidating them into two processes cuts both latency and CPU.

Measured on an `Edit` event, spawning the actual `hooks.json` commands in parallel like the harness does (median of 7 runs):

| | main | this PR |
|---|---|---|
| Blocking processes per tool call | 7 | 1 |
| Blocking hook latency | 81 ms | 49 ms (~40% faster) |
| Background processes per tool call | 3 | 1 |

Note: `observe-runner` now runs in-process (~370 ms) inside the async dispatcher, but that dispatcher stays backgrounded, so it adds no user-facing latency.

## Testing Done

- [x] Manual testing completed — piped real `PostToolUse` payloads through both dispatchers: console.log warning fires on stderr, stdout pass-through is byte-identical, `ECC_DISABLED_HOOKS=post:edit:console-warn` and `ECC_HOOK_PROFILE=minimal` still gate correctly
- [x] Automated tests pass locally (`node tests/run-all.js`) — 3116/3116
- [x] Edge cases considered and tested — truncated stdin (>1MB suppresses pass-through), multi-byte UTF-8 across chunk boundaries, per-hook failure isolation with exit-code propagation

## Type of Change
- [x] `refactor:` Code refactoring

## Security & Quality Checklist
- [x] No secrets or API keys committed (ghp_, sk-, AKIA, xoxb, xoxp patterns checked)
- [x] JSON files validate cleanly
- [x] Shell scripts pass shellcheck (if applicable)
- [x] Pre-commit hooks pass locally (if configured)
- [x] No sensitive data exposed in logs or output
- [x] Follows conventional commits format

## If you changed dependencies or `package.json` (`bin` / `files` / deps)
- [x] Not applicable — dependencies and package.json were not changed.

## If you added a skill, command, agent, hook, or CLI tool
- [x] New script path `scripts/hooks/posttooluse-dispatcher.js` is covered by the publish surface (`tests/scripts/npm-publish-surface.test.js` passes); no new skill/command/agent, so catalog/registry/docs tables unchanged

## Documentation
- [x] Updated relevant documentation (`commands/quality-gate.md` hook wiring notes)
- [x] Added comments for complex logic
- [x] README updated (if needed) — not needed

### CODE REVIEW

File: scripts/hooks/posttooluse-dispatcher.js [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+  if (stdout) process.stdout.write(stdout);
+  process.exitCode = result.exitCode;
+}
+
+if (require.main === module || require.main === undefined) {
+  main().catch(error => {
+    process.stderr.write(`[Hook] PostToolUse dispatcher failed: ${error.message}\n`);
+    process.exitCode = 0;
+  });
+}
```
[the remark this answers is not in the document]
Comment (pythonstrup) [reply scripts/hooks/posttooluse-dispatcher.js]: Fixed in fc489a83 — replaced the guard with an explicit exported `cli()` and the `hooks.json` bootstraps now call `require(s).cli()`. The module-level guard is back to the standard `require.main === module` only, so requiring the dispatcher (dashboard-web, test runners, Jest, worker threads) can never trigger dispatch. Added a regression test that requires the module with a payload on stdin and asserts nothing is dispatched.

File: scripts/hooks/posttooluse-dispatcher.js [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+
+    try {
+      const result = normalizeResult(raw, hook.run(raw, {
+        hookId: hook.id,
+        pluginRoot,
+        scriptPath: path.join(pluginRoot, hook.script || ''),
+        truncated: options.truncated === true,
+        maxStdin: MAX_STDIN,
+      }));
+      if (result.stdout) stdout = result.stdout;
```
[the remark this answers is not in the document]
Comment (pythonstrup) [reply scripts/hooks/posttooluse-dispatcher.js]: Fixed in fc489a83 — added `mergeHookStdout()`: multiple `additionalContext` envelopes now merge into a single PostToolUse envelope (contexts joined with newline); when a raw stdout can't be merged, the last hook's output wins and a stderr warning names the dropped hook IDs, so nothing is lost silently. Covered by a new unit test.

**Comment by pythonstrup** [comment]
Re: the P1 `post:bash:dispatcher silently loses profile-independence` — partially disagree with the premise, but hardened anyway in 83ba914b.

The claim that profile-less sub-hooks (`post:bash:command-log-audit`, `post:bash:command-log-cost`) ran in all profiles including minimal is incorrect: `parseProfiles()` in `scripts/lib/hook-flags.js` defaults unspecified profiles to `['standard', 'strict']`, and `bash-hook-dispatcher.js` gates every sub-hook through `isHookEnabled()`. Under `ECC_HOOK_PROFILE=minimal`, main spawned the bash dispatcher process but skipped every sub-hook — minimal users never received audit/cost log entries, so no user-visible regression was introduced.

That said, main's semantic was "always run the phase, let sub-hooks gate themselves," so the outer registry gate is now `minimal,standard,strict` to restore that exactly (a future sub-hook opting into minimal won't be blocked at the phase level), plus the previously missing minimal-profile async dry-run regression test.

File: tests/hooks/posttooluse-dispatcher.test.js [hunk 3]
```diff
@@ -0,0 +1,434 @@
...
+      assert.ok(result.stderr.includes('post:test:broken'));
+      assert.ok(result.stderr.includes('boom'));
+      assert.ok(result.stderr.includes('post:test:nonzero'));
+      assert.ok(result.stderr.indexOf('post:test:broken') < result.stderr.indexOf('last warning'));
+      assert.strictEqual(result.exitCode, 7, 'explicit child exit codes should be preserved');
+      assert.strictEqual(resolveMainStdout(raw, { stdout: '', exitCode: 7 }, { passthrough: true, truncated: false }), '', 'nonzero results should not restore raw input');
+    })
+  )
+    passed++;
+  else failed++;
```
[the remark this answers is not in the document]
Comment (pythonstrup) [reply tests/hooks/posttooluse-dispatcher.test.js:344]: Added in 4338479a — new test spawns the real dispatcher subprocess with an injected failing hook (mutating the exported SYNC_HOOKS registry, then calling cli()) and asserts the OS-level `result.status === 7`, the stderr diagnostic, and that pass-through output stays suppressed on failure.

File: scripts/hooks/post-edit-console-warn.js [hunk 4]
```diff
@@ -12,43 +12,54 @@
...
-
-process.stdin.on('end', () => {
+function run(data) {
+  const warnings = [];
   try {
     const input = JSON.parse(data);
     const filePath = input.tool_input?.file_path;
 
     if (filePath && /\.(ts|tsx|js|jsx)$/.test(filePath)) {
       const content = readFile(filePath);
```
[the remark this answers is not in the document]
Comment (pythonstrup) [reply scripts/hooks/post-edit-console-warn.js:22]: Skipping this one intentionally. The hook's input is the PostToolUse payload from the local Claude Code harness — the same trust domain as the user's own shell — not remote data. Its only output is up to five console.log-matching lines written to the user's local stderr, so there is no exfiltration path. Constraining `file_path` to the repo root would break the hook's actual purpose: it must warn on files the user edits anywhere (other projects, dotfiles, monorepo siblings), and this behavior is unchanged from main (the refactor only moved the existing logic into an exported run()). Happy to revisit if the hook ever starts echoing file contents beyond matched lines or writing anywhere other than local stderr.

**Review by daltino — APPROVED** [review]
This PR consolidates PostToolUse hooks into unified sync/async dispatchers, streamlining the hook system and improving maintainability. The changes remove redundant code, centralize logic into a new `posttooluse-dispatcher.js` script, and update affected files like `hooks.json` and related scripts. The refactor is clean, aligns with the contribution guidelines for enhancing hooks, and simplifies hook dispatching effectively. Good work!

**Review by mc856 — APPROVED** [review]
Ran an A/B against main (node 22): same Edit payload through the sync dispatcher — stderr warnings byte-identical, echo/exit parity, ECC_DISABLED_HOOKS and profile gating unchanged; bash async path likewise. Checked all 10 PostToolUse registrations for matcher/profile/timeout equivalence — they hold. Serializing metrics-bridge before context-monitor is also a quiet fix: on main those two race on the bridge file, and I could reproduce the ordered dataflow (bridge → monitor loop-warning envelope) working deterministically here.

One cross-PR note re #2485: that PR removes the raw-payload echo ("empty stdout + exit 0 = no opinion"), citing per-tool-call PostToolUse-sized payloads as the main transcript-bloat source. This PR routes PostToolUse around run-with-flags entirely and re-pins the echo via `ECC_POSTTOOLUSE_PASSTHROUGH='1'` in the two dispatcher commands — right for behavior parity with today's main, but if #2485 lands, whichever merges second should drop that env assignment (one line; the dispatcher already treats unset as no-echo). The two merge cleanly at the git level, so this won't surface as a conflict — flagging so it doesn't get lost.
